dpnp/backend/extensions/blas/dot.cpp

```diff
@@ -136,88 +136,97 @@ std::pair<sycl::event, sycl::event> dot(sycl::queue &exec_q,
+    if (result_typenum != vectorX_typenum || result_typenum != vectorY_typenum)
+    {
+        throw py::value_error("Given arrays must be of the same type.");
     }
 
     auto array_types = dpctl_td_ns::usm_ndarray_types();
-    int vectorAB_type_id = array_types.typenum_to_lookup_id(vectorA_typenum);
-    int result_type_id = array_types.typenum_to_lookup_id(result_typenum);
+    int type_id = array_types.typenum_to_lookup_id(vectorX_typenum);
 
-    dot_impl_fn_ptr_t dot_fn =
-        dot_dispatch_table[vectorAB_type_id][result_type_id];
+    dot_impl_fn_ptr_t dot_fn = dot_dispatch_table[type_id][type_id];
     if (dot_fn == nullptr) {
         throw py::value_error(
             "Types of input vectors and result array are mismatched.");
     }
 
-    char *a_typeless_ptr = vectorA.get_data();
-    char *b_typeless_ptr = vectorB.get_data();
+    char *x_typeless_ptr = vectorX.get_data();
+    char *y_typeless_ptr = vectorY.get_data();
     char *r_typeless_ptr = result.get_data();
 
-    const int a_elemsize = vectorA.get_elemsize();
-    const int b_elemsize = vectorB.get_elemsize();
-    if (str_a < 0) {
-        a_typeless_ptr -= (n - 1) * std::abs(str_a) * a_elemsize;
+    std::vector<py::ssize_t> x_stride = vectorX.get_strides_vector();
+    std::vector<py::ssize_t> y_stride = vectorY.get_strides_vector();
+    const int x_elemsize = vectorX.get_elemsize();
+    const int y_elemsize = vectorY.get_elemsize();
+
+    const std::int64_t incx = x_stride[0];
+    const std::int64_t incy = y_stride[0];
+    // In OneMKL, the pointer should always point out to the first element of
+    // the array and OneMKL handle the rest depending on the sign of stride.
+    // In OneMKL, when the stride is positive, the data is read in order and
+    // when it is negative, the data is read in reverse order while pointer
+    // always point to the first element
+    // When the stride is negative, the pointer of the array coming from dpnp
+    // points to the last element. So, we need to adjust the pointer
+    if (incx < 0) {
+        x_typeless_ptr -= (n - 1) * std::abs(incx) * x_elemsize;
     }
-    if (str_b < 0) {
-        b_typeless_ptr -= (n - 1) * std::abs(str_b) * b_elemsize;
+    if (incy < 0) {
+        y_typeless_ptr -= (n - 1) * std::abs(incy) * y_elemsize;
     }
 
-    sycl::event dot_ev = dot_fn(exec_q, n, a_typeless_ptr, str_a,
-                                b_typeless_ptr, str_b, r_typeless_ptr, depends);
+    sycl::event dot_ev = dot_fn(exec_q, n, x_typeless_ptr, incx, y_typeless_ptr,
+                                incy, r_typeless_ptr, depends);
 
     sycl::event args_ev = dpctl::utils::keep_args_alive(
-        exec_q, {vectorA, vectorB, result}, {dot_ev});
+        exec_q, {vectorX, vectorY, result}, {dot_ev});
 
     return std::make_pair(args_ev, dot_ev);
 }
 
-template <typename fnT, typename Tab, typename Tc>
+template <typename fnT, typename varT, typename none>
```

Comment:
After my suggestion you can change this to `template <typename fnT, typename T>`